File: editors/intellij/src/test/kotlin/com/mediusfy/modulex/intellij/core/RealServerE2ETest.kt
package com.mediusfy.modulex.intellij.core

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import java.io.File
import kotlin.test.Test
import kotlin.test.assertTrue
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable

/**
 * Drives the real tools/mcpserver (spawned via go run) with the plugin's
 * MCP client, mirroring the validation done for editors/vscode. Opt-in
 * (MODULEX_E2E=1 and a modulex checkout two levels up) because it builds
 * and runs the Go server.
 */
@EnabledIfEnvironmentVariable(named = "MODULEX_E2E", matches = "1")
class RealServerE2ETest {
    @Test
    fun realServerRoundTrip() {
        val root = File("../..").canonicalFile
        assertTrue(root.resolve("tools/mcpserver").isDirectory, "not a modulex checkout: $root")
        val spec = resolveServerSpec(root, emptyList())!!
        val client = McpClient.start(spec) { }
        try {
            val discover = JsonObject()
            discover.addProperty("root", root.path)
            val repo = client.callTool(Tool.DISCOVER_REPOSITORY, discover)
            assertTrue(repo.has("repository"), "discover_repository returned no repository")

            val recommend = JsonObject()
            recommend.add("changed_files", JsonArray())
            val plan =
                com.google.gson.Gson()
                    .fromJson(
                        client.callTool(Tool.RECOMMEND_VERIFICATION, recommend).getAsJsonObject("plan"),
                        Plan::class.java,
                    )
            assertTrue(plan.fullGates.isNotEmpty(), "expected full gates in the plan")

            val review = JsonObject()
            review.addProperty("root", root.path)
            review.addProperty("base_ref", "HEAD")
            review.addProperty("head_ref", "HEAD")
            val results =
                client.callTool(Tool.REVIEW_DIFF, review).getAsJsonArray("results")
            assertTrue(results.size() > 0, "expected review results")
            // check-changelog and friends judge the branch against
            // origin/main regardless of base_ref, so their verdict depends
            // on the checkout's state; assert the protocol round-trip
            // (well-formed results), not a particular verdict.
            val knownStatuses = setOf("pass", "fail", "skipped", "unavailable", "approval_required")
            for (r in results) {
                val obj = r.asJsonObject
                assertTrue(obj.get("name").asString.isNotEmpty(), "result missing name")
                assertTrue(
                    obj.get("status").asString in knownStatuses,
                    "check ${obj.get("name").asString} has unknown status ${obj.get("status").asString}",
                )
            }
        } finally {
            client.dispose()
        }
    }
}

File: editors/vscode/test/realserver.e2e.test.ts
import assert from "node:assert/strict";
import * as fs from "node:fs";
import * as path from "node:path";
import test from "node:test";
import { McpClient } from "../src/mcpClient";
import { resolveServerSpec } from "../src/server";

// Opt-in end-to-end test against the real Go MCP server (mirrors the
// IntelliJ RealServerE2ETest): spawns tools/mcpserver via go run from the
// repository root and drives it with the extension's own client. Run with:
//
//   MODULEX_E2E=1 npm test
//
// Skipped otherwise, because it builds and runs the Go server.
// __dirname is out/test at run time (tests run compiled), so the repo root
// is four levels up: out/test -> vscode -> editors -> repo.
const repoRoot = path.resolve(__dirname, "..", "..", "..", "..");
const enabled =
  process.env.MODULEX_E2E === "1" &&
  fs.existsSync(path.join(repoRoot, "tools", "mcpserver"));

test("real server round-trip", { skip: !enabled }, async () => {
  const spec = resolveServerSpec(repoRoot, []);
  assert.ok(spec, "modulex checkout not auto-detected");
  const client = await McpClient.start(spec, () => {});
  try {
    const disc = await client.callTool<{ repository: object }>("discover_repository", {
      root: repoRoot,
    });
    assert.ok(disc.repository, "discover_repository returned no repository");

    const rec = await client.callTool<{ plan: { FullGates: unknown[] } }>(
      "recommend_verification",
      { changed_files: [] },
    );
    assert.ok(rec.plan.FullGates.length > 0, "expected full gates in the plan");

    const review = await client.callTool<{
      results: Array<{ name: string; status: string }>;
    }>("review_diff", { root: repoRoot, base_ref: "HEAD", head_ref: "HEAD" });
    assert.ok(review.results.length > 0, "expected review results");
    const known = new Set(["pass", "fail", "skipped", "unavailable", "approval_required"]);
    for (const r of review.results) {
      assert.ok(known.has(r.status), `check ${r.name} has unknown status ${r.status}`);
    }
  } finally {
    client.dispose();
  }
});
